worker/restore_reduce.go

```diff
@@ -0,0 +1,314 @@
+			x.FixedDuration(since), humanize.IBytes(read), humanize.IBytes(proc),
+			humanize.IBytes(pr), humanize.IBytes(uint64(z.NumAllocBytes())))
+	}
+	for {
+		select {
+		case <-closer.HasBeenClosed():
+			update()
+			glog.Infof("Restore REDUCE Done in %s.\n", x.FixedDuration(time.Since(start)))
+			return
+		case <-ticker.C:
+			update()
+		}
+	}
+}
+
+func (r *reducer) Reduce() error {
+	var files []string
+
+	var total int64
+	f := func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if strings.HasSuffix(info.Name(), ".map") {
+			files = append(files, path)
+			total += info.Size()
+		}
+		return nil
+	}
+
+	if err := filepath.Walk(r.mapDir, f); err != nil {
+		return err
+	}
+	glog.Infof("Got %d map files of compressed size: %s.\n",
+		len(files), humanize.IBytes(uint64(total)))
+
+	// Pick up map iterators and partition keys.
+	partitions := make(map[string]struct{})
+	for _, fname := range files {
+		header, itr := newMapIterator(fname)
+		for _, k := range header.PartitionKeys {
+			if len(k) == 0 {
+				continue
+			}
+			partitions[string(k)] = struct{}{}
+		}
+		r.mapItrs = append(r.mapItrs, itr)
+	}
+
+	keys := make([][]byte, 0, len(partitions))
+	for k := range partitions {
+		keys = append(keys, []byte(k))
+	}
+	sort.Slice(keys, func(i, j int) bool {
+		return y.CompareKeys(keys[i], keys[j]) < 0
+	})
+	// Append nil for the last entries.
+	keys = append(keys, nil)
+	r.partitionKeys = keys
+
+	errCh := make(chan error, 2)
+	go func() {
+		errCh <- r.blockingRead()
+	}()
+	go func() {
+		errCh <- r.process()
+	}()
+
+	for i := 0; i < 2; i++ {
+		if err := <-errCh; err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *reducer) blockingRead() error {
+	cbuf := z.NewBuffer(64<<20, "Restore.GetBuf")
+
+	sortAndPush := func(buf *z.Buffer) {
+		// Let's sort here. So, there's less work for processor.
+		buf.SortSlice(func(ls, rs []byte) bool {
+			lme := mapEntry(ls)
+			rme := mapEntry(rs)
+			return y.CompareKeys(lme.Key(), rme.Key()) < 0
+		})
+		atomic.AddUint64(&r.bytesRead, uint64(buf.LenNoPadding()))
+		r.bufferCh <- buf
+	}
+	for _, pkey := range r.partitionKeys {
+		for _, itr := range r.mapItrs {
+			if err := itr.Next(cbuf, pkey); err != nil {
+				cbuf.Release()
+				return err
+			}
+		}
+		if cbuf.LenNoPadding() < 256<<20 {
+			// Pick up more data.
+			continue
+		}
+		sortAndPush(cbuf)
+		cbuf = z.NewBuffer(64<<20, "Restore.GetBuf")
+	}
+
+	if !cbuf.IsEmpty() {
+		sortAndPush(cbuf)
+	} else {
+		cbuf.Release()
+	}
+	close(r.bufferCh)
+	return nil
+}
+
+func (r *reducer) process() error {
+	if r.w == nil {
+		return nil
+	}
+	writer := r.w
+
+	kvBuf := z.NewBuffer(64<<20, "Restore.GetBuf")
+	defer func() {
+		kvBuf.Release()
+	}()
+
+	var lastKey []byte
+	for cbuf := range r.bufferCh {
+		err := cbuf.SliceIterate(func(s []byte) error {
+			me := mapEntry(s)
+			key := me.Key()
+
+			// Don't need to pick multiple versions of the same key.
+			if y.SameKey(key, lastKey) {
```

Comment:
Any reason we would get the same key again?

Reply:
It could come from different backups. We are trying to restore from more than one backups in one shot.